Mathlib/Order/FunLike.lean

```diff
@@ -0,0 +1,48 @@
+/-
+Copyright (c) 2024 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+import Mathlib.Data.FunLike.Basic
+import Mathlib.Order.Basic
+import Mathlib.Tactic.GCongr.Core
+
+/-!
+# Preorder and partial order on bundled homs
+
+In this file we define `LE`, `Preorder`, and `PartialOrder` instances
+on any type with `DFunLike` instance and appropriate instances on the codomains.
+-/
+
+namespace DFunLike
+
+variable {F : Type*} {α : Type*} {β : α → Type*} [DFunLike F α β]
+
+section LE
+
+variable (F) in
+/-- A mixin typeclass saying that the order on bundled morphisms
+agrees with the pointwise order on corresponding functions. -/
+class PointwiseLE [∀ a, LE (β a)] [LE F] : Prop where
+  le_iff : ∀ {f g : F}, f ≤ g ↔ ∀ a, f a ≤ g a := by intro; rfl
+
+variable [∀ a, LE (β a)] [LE F] [PointwiseLE F] {f g : F}
+
+theorem le_def : f ≤ g ↔ ∀ a, f a ≤ g a := PointwiseLE.le_iff
+
+@[simp, norm_cast] theorem coe_le_coe : coe f ≤ g ↔ f ≤ g := le_def.symm
+
+@[gcongr] theorem apply_mono (h : f ≤ g) (x : α) : f x ≤ g x := le_def.1 h x
```

Comment:
Isn't that a terrible `gcongr` lemma performance-wise? Can you try an example of the form
```
example : f1 (f2 (f3 x))) ≤ g1 (g2 (g3 x))) := by gcongr
```
(with suitable typeclass assumptions) and see whether heartbeats grow linearly or exponentially in the number of functions?

Reply:
I removed the `@[gcongr]` tag for now.